New Mech UI and equipment refactor

MTandi opened the pull request:
![bWJlVIDO53](https://github.com/tgstation/tgstation/assets/3625094/d75030b5-59e9-43f6-96b4-1b7564caceef)

## About The Pull Request

Made a new UI and refactored some mech code in the process.

Fixes #66048
Fixes #77051
Fixes #65958 ??? if it was broken
Fixes #73051 - see details below
Fixes other undocumented things, see changelog.

## Why It's Good For The Game

The UI was too bulky and Mechs were too complex for no reason. 
Now they follow some general rules shared between other SS13 machinery, and there is less magic happening under the hood.

## Detailed Changes

### New Mech UI, Air Tank and Radio as separate modules

Previous UI for comparison:

<img alt="9SScrXAOjy" src="https://github.com/tgstation/tgstation/assets/3625094/904e3d07-e7d7-4d3a-a062-93e0e35b4b66">

Previously mechs came with radio pre-installed and air canisters magically pre-filled with air even when you build one in fab.
Radio and Air Tanks are now both utility modules that are optional to install. Gas RCS thrusters still require Air Tank module to operate.

This made the Mechs more barebones when built, giving you only the basic functionality.

<img alt="5SDMlXTJxv" src="https://github.com/tgstation/tgstation/assets/3625094/b9364230-49ac-416b-aa70-e851fbf2050e">

To compensate this change, all mechs got two extra utility module slots.

All other modules got new UI. And ore box now shows the list of ores inside.

<img alt="SRX5FjvsHA" src="https://github.com/tgstation/tgstation/assets/3625094/cbe2e98d-1cd4-4667-8dae-2f9227b4b265">

### Mounted Radio

Works as a normal radio, but with subspace transmission. Available from the basic mech research node and can be printed in fab.

### Cabin Sealing

To compensate for the lack of air tank by default, mechs with enclosed cabin (e.g. all except Ripley) got an ability to toggle cabin exposure to the outside air. Exiting the mech makes cabin air automatically exposed. 

When you seal the cabin, it traps some of the outside air inside the cabin and you can breathe with this air to perform a short space trips. But the oxygen will run out quickly and CO2 will build up. 

Sealing the cabin in space will make the cabin filled with vacuum, and it will stay there until you return to air environment and unseal the cabin, letting the breathable air to enter. There are temperature and pressure sensors that turn yellow/red when the corresponding warning thresholds are reached.

You could also use personal internals in combination with cabin sealing for long space travels, so Air Tank is completely optional now and mostly needed when you need RCS thruster.

### RCS thrusters

They are now available earlier in tech tree and consume reasonable amount of air (5 times more than human jetpacks), and they don't work without Mounted Air Tank, unless it's an Ion thruster variant.

### Mounted Air Tank

Available from the basic mech research node and can be printed in fab. Built model comes empty, and syndicate mechs come with one full of oxygen.

<img alt="GrFDrH5Hwe" src="https://github.com/tgstation/tgstation/assets/3625094/b677b705-bda0-4c8c-96c7-d32bf7bf9f28">

Can be switched to pressurize or not pressurize the cabin. Releases gas only when the cabin is sealed shut. Starts releasing automatically, but can be toggled to not release if you want to use it just as a portable canister. 

Cabin pressure can now be configured in the module UI instead of Maintenance UI.

Can be attached to a pipe network when the mech is parked above a connection port.

Comes with a pump that works similarly to the portable pump. It lets you vent the air tank contents outside, or suck air from the room to fill the air tank. Intended to provide an ability to fill the air tank without the need to bother with pipes.

Also has gas sensors that display gas mix data of the tank and the cabin (when sealed).

### Stock part changes

All mechs now require a servo motor and they reduce mech movement power consumption instead of scanning module.

Scanning modules are optional for mech operation (still required to build) and the lack of one disables the following UI elements:

- Display of mech integrity (you can still see the alerts or examine the mech to get rough idea)
- Display of mech status on internal damage (and you can't repair what you can't diagnose)

The rating of scanning module doesn't have any effect as of now.

Cargo mech comes without it roundstart.

<img alt="2vDtt99oqb" src="https://github.com/tgstation/tgstation/assets/3625094/147144ca-824e-4501-acf5-6ee104f309e7">

Capacitors now also reduce light power usage and raise the overclocking temperature thresholds (see below).

### Maintenance

Maintenance UI removed, and its logic migrated to other places.

Access modification now managed inside the mech, and anyone who can control the mech, can adjust the access in the same way as they can set DNA key.

To open the maintenance panel you just need a screwdriver. It is instant when the mech is empty and it has a 5 second delay when there is an occupant to avoid in-combat hacking and part removal. It will alert the occupant that someone is trying to tinker with their mech.

![image](https://github.com/tgstation/tgstation/assets/3625094/1abfca3c-8ba9-44b0-913c-c209564b91fd)

Once the panel is open, you can see the part ratings:

![image](https://github.com/tgstation/tgstation/assets/3625094/404f95bb-9f74-4e5b-a975-5ab6f74bdfa9)

With open panel you can hack the mech wires (roboticists can now see them):

<img alt="mj205G2qDa" src="https://github.com/tgstation/tgstation/assets/3625094/44cea0d1-44b4-4b50-b1d3-a97c0056bab3">

There are wires for:
- Enabling/Disabling ID and DNA locks
- Toggling mech lights
- Toggling mech circuits malfunction (battery drain, sparks) 
- Toggling mech equipment malfunction (to repair after EMP or cause EMP-like effect, disarming mech)
- 3 dud wires that do nothing

The hacker may be shocked if the mech power cell allows.

When the panel is open and the user has access to the mech, they can remove parts with a crowbar:

<img alt="jR40gyTWtJ" src="https://github.com/tgstation/tgstation/assets/3625094/b573f5b9-b8ea-412e-b3e0-c872e01e0c23">

Hitting the mech with an ID from outside now toggles the ID Lock on/off if the ID has sufficient access.

### Power consumption and overclocking

Rebalanced mech power consumption. T4 parts were not working in Syndicate Mechs, as their effect was not calculated until you manipulate parts manually. Constructed mechs with t1 parts even had their energy drain reduced with upgrade to t1.

Now all mechs apply their base step power usage correctly don't ignore the stock parts.

Servo tier now reduces base power consumption by 0% at t1, 50% at t2, 33% at t3 and 25% at t4
Capacitor tier now reduces base power consumption of mele attacks, phasing and light by the same amounts.

Gygax leg actuators overload replaced with mech overclocking. Any mech can be overclocked by hacking wires, but only Gygax has a button for toggling it from the Cabin.

Now there is an overclock coefficient. 1.5 for Gygax and other mechs, 2 for Dark Gygax.

When overclocked, mechs moves N times faster, but consumes N times more power.

![image](https://github.com/tgstation/tgstation/assets/3625094/01e285fd-6fd6-4558-8277-2ed3abf474d8)

While overclocked, mech heats up every second, regardless of movement, and starts receiving internal and integrity damage after a certain temperature threshold. The chance is 0% at the threshold, and 100% at thresholds * 2. The roll happens every tick. Capacitor upgrades this threshold, letting you overclock safely for longer periods.

![image](https://github.com/tgstation/tgstation/assets/3625094/80d90cea-0d20-4054-9369-a47deb6f52f2)

When you stop overclock, the temperature goes back down.

### Other changes and fixes

Concealed weapon bay now doesn't show up when you examine the mech, so it's actually concealed now.

New radio module can properly change its frequency, as it didn't work for previous radio.

Launcher type weapons were ignoring cooldowns and power usage, so you could spam explosive banana peals, while they should have a 2 second cooldown:
<img alt="q5GjUsHwGr" src="https://github.com/tgstation/tgstation/assets/3625094/d102725d-e9e1-4588-9d6c-b9e38b7a6535">

Now this is fixed and all launcher type weapons properly use power and have their cooldowns working.
And now they have the kickback effect working (when it pushes you in the opposite direction in zero gravity on throw).

Thermoregulator now heats/cools considering heat capacity instead of adding/reducing flat 10 degrees. So you can heat up cabin air quicker if the pressure is low.

There were some other sloppy mistakes in mech code, like some functions returning too early, blocking other functionality unintentionally. Fixed these and made some other minor changes and improvements.

## Changelog

:cl:
refactor: Refactored Mech UI
refactor: Refactored mech radio into a utility module, adding extra slot to all mechs
refactor: Refactored mech air tank into a utility module with an air pump, adding extra slot to all mechs
refactor: Refactored mech cabin air - there is now a button to seal or unseal cabin to make it airtight or exchanging gases with the environment
refactor: Removed mech maintenance UI Access is set in mech UI, and parts are ejected with a crowbar
add: Mech now has wires and can be hacked
qol: Roboticists now can see MOD suit and mech wires
add: Mechs now require servo motor stock part and it affects movement power usage instead of scanning module
add: Scanning module absence doesnt block mech movement and hides some UI data instead. Big Bess starts without one.
qol: Hitting mech with ID card now toggles ID lock on/off if the card has required access
fix: Fixed concealed weapon bay not being concealed on mech examine
fix: Fixed mech radio not changing frequency
fix: Fixed mech launcher type weapons ignoring specified cooldown
fix: Fixed mech launcher type weapons not using specified power amount
fix: Fixed mech temperature regulator ignoring gas heat capacity
fix: Fixed mech stopping processing other things while not heating internal air
fix: Fixed mech being able to leave transit tube in transit
fix: Fixed mech internal damage flags working incorrectly
fix: Fixed Gygax leg overloading being useless
fix: Fixed mechs ignoring their stock parts on creation. Syndicate mechs now stronger against lasers and consume less energy on move. Upgrading from tier 1 to tier 2 doesn't make mech consume MORE energy than before the upgrade.
balance: Rebalanced mech energy drain with part upgrades. Base energy drain reduced by 50%, 33%, 25% with upgrades and applies to movement (Servo rating), phasing, punching, light (Capacitor rating).
balance: Hydraulic clamp now can force open airlocks
balance: Made mech RCS pack consume reasonable amount of gas
code: Fixed some other minor bugs and made some minor changes in the mech code
/:cl:

MTandi committed:
draft

MTandi committed:
Merge branch 'master' into mechui

MTandi committed:
utilities refactored

MTandi committed:
compact ui

MTandi committed:
UI tweaks

MTandi committed:
usage of new air tank

MTandi committed:
ui_theme, radio, air fix

MTandi committed:
cleanup

MTandi committed:
more cleanup

MTandi committed:
thrusters fix

MTandi committed:
selected module logic

MTandi committed:
icons for new modules

MTandi committed:
syringe gun sprite

MTandi committed:
tweaks

MTandi committed:
sealed cabin and air tank working

MTandi committed:
smaller tank

MTandi committed:
orebox, cargo have separate ui

MTandi committed:
air pump, alerts, pressures, fixes

MTandi committed:
ore box ui fix, pump tweaks

MTandi committed:
maintenance removed

MTandi committed:
Merge branch 'master' into mechui

MTandi committed:
UI cleanup

MTandi committed:
cleanup

MTandi committed:
hacking, scanmod stuff

MTandi committed:
examine fix

atlasle commented:
this is fantastic man you're on a roll

CRITAWAKETS commented:
by the way, /obj/item/mecha_parts/mecha_equipment/Destroy() is broken and causes mech equipment destruction to not work. It's a simple fix, just make the item detach after doing the to_chat and sound rather than before.

MTandi committed:
forgot qdell

MTandi commented:
> by the way, /obj/item/mecha_parts/mecha_equipment/Destroy() is broken and causes mech equipment destruction to not work. It's a simple fix, just make the item detach after doing the to_chat and sound rather than before.

I'm planning to redo mecha damage to make equipment shield damage, losing integrity and falling off the mech in the process. Didn't want to bundle combat balance stuff into this PR. I'll take a look into this.

MTandi committed:
icon fix

MTandi committed:
sleeper icon fix

githubuser4141 commented:
holy shit this is fire, but does this mean having an unsealed cabin in a seal-ready mech means you fall out if you're incapacitated? or even if there's an air malfunction? Seems kinda odd because no seals would probably just be a small hole to allow air to be exchanged rather then the entire hatch being open right?

Sealed101 commented:
you should elaborate on the changelog more in the PR body, because there's quite a lot of significant changes to the things affected. 
- what does the scanning module do now, since the servo now affects power usage? 
- how does one access the maintenance panel now? 
- what do the wires of the mech do? 
- how do the air tank and the radio modules operate, since they seem to be also printable from the fab? 
- there's a kickback thing in there somewhere, what does that apply to? 
- etc.

MTandi committed:
better wording

MTandi committed:
delayed screwdriving, roboticists see wires

MTandi committed:
wording

MTandi commented:
> holy shit this is fire, but does this mean having an unsealed cabin in a seal-ready mech means you fall out if you're incapacitated? or even if there's an air malfunction? Seems kinda odd because no seals would probably just be a small hole to allow air to be exchanged rather then the entire hatch being open right?

It's a different variable, so `cabin_sealed` is unrelated to `enclosed`. You guessed right, opening seal doesn't make you visible from outside or fall out of the mech when incapacitated. I changed the wording in UI to not confuse people.

> This UI can still be moved around the screen and you can interact with your mech while it's open, right? Could a feature that lets 'mech rentals' be feasible be added, a way for the original owner to DNA-lock their mech, then have a lock settings menu that allows unauthorized drivers to access, but not modify the mech? I think it'd encourage people to give their mechs away to security/mining/random assistants because it'd be some assurance it'd be used the way it was intended to be, and the person(s) wouldn't DNA lock it for themselves.

I simplified mech access to make it more accessible, removing the ability to disable access modification. My assumption is that if someone entered the mech, it is their as they can set DNA lock, or destroy the mech due to their actions. But to compensate for this, now  as a roboticists that rents the mech, you can hack it in case if they change access or set DNA lock and abandon the mech - you will see the wire names.

@Sealed101 Updated the PR body

MTandi committed:
equip destroy fix

atlasle commented:
now if only mechs could be actually useful

MTandi commented:
> now if only mechs could be actually useful

The concept is valid, the rest is a matter of balance and providing more tools for different use cases.

MTandi committed:
dark honker syndicate theme

Sealed101 commented:
wait this uses ui themes now
the perfect moment to restore honker's previous fugly rainbow UI has come
pls 🙏

MTandi committed:
mechs can't leave transit tubes in transit

MTandi committed:
honker yellow ui

MTandi commented:
> wait this uses ui themes now the perfect moment to restore honker's previous fugly rainbow UI has come pls 🙏

Changed it to this to look like it's made of bananium:

![image](https://github.com/tgstation/tgstation/assets/3625094/593a56fb-8a54-417f-a1be-45aae7280776)

atlasle commented:
> The concept is valid, the rest is a matter of balance and providing more tools for different use cases.

absolutely agree, mechs are a sci fi concept we should never get rid off, and i agree that they definitely need some more ways to either subvert their drawbacks or define even more their advantages

moocowswag commented:
Mechs have plenty of tools outside of robotics and I dont like that people think that what you can directly put on them is the only strength that they can collect.

A roboticist easily has access to the circuit room, which means he has access to signalers that he can activate while in a mech, now all of a sudden you have the possibility of extra (oh shit) buttons you can use with chem grenades within your own inventory or a launchpad that teleports your mech to safety


You can even do simpler stuff like just putting a quantum spin inverter in your mech and linking it to another one that someone else has, then asking them to activate it when needed.

Helg2 commented:
Something NT themed for mechs like Seraph which is NT mech iirc would be cool (despite it never being used, but still)

atlasle commented:
> This is in response to the comments about the current state of mech balance (alittle off topic but still relevant)

The problem is that all of these things can be done better for a bigger payout when not in a mech

MTandi commented:
> Something NT themed for mechs like Seraph which is NT mech iirc would be cool (despite it never being used, but still) And also the mimes mech perhaps?

NTOS is the default theme for mechs already.

@githubuser4141 this is actually my mistake, I wanted to replace enclosed with cabin_sealed, but then abandoned this idea and forgot about this line. Will revert it back to check for only `!enclosed`. Cabin can't become sealed when it's not enclosed.

>PS: Does the UI support more then one tool on an arm? E.g. if someone wanted to make a super mech with 2 equipments on the right arm, would that work?

The UI now does, but there is no system to switch which of the left arm slots will be triggered with LMB. I made the UI compatible with this, but I plan to add a 4-armed mech later and support the switching on backend (no promises though).

@moocowswag players like you will find a way to have fun or abuse anything in the game as all mechanics we have are with loopholes. But mechs have the same issue as circuits - it's hard to justify using one for your task, like "To do this, having a mech would be helpful". Mechs should be more accessible and easy to get on with for the general audience.

MTandi committed:
falling out only when not enclosed

MTandi committed:
bring the icon back here

MTandi committed:
order

MTandi committed:
wrong note

githubuser4141 commented:
Would it be possible to explain mech armour dynamics (directional armour, internal threshold minimums, the fact that AP doesn't count) in the UI? Currently, absolutely none of the former are shown or explained *anywhere* in-game. You'll not find a single in-game source that even mentions mecha directional armour or damage thresholds despite these mechanics being hugely important in a lot of fights.

Sealed101 commented:
> Would it be possible to explain mech armour dynamics (directional armour, internal threshold minimums) in the UI? Currently, absolutely none of the former are shown or explained _anywhere_ in-game.

brother the game itself doesn't know half the time what mech armor does
- #74022

MTandi commented:
> Would it be possible to explain mech armour dynamics (directional armour, internal threshold minimums) in the UI? Currently, absolutely none of the former are shown or explained _anywhere_ in-game.

I want to rework mechs combat making their total integrity depend on modules, making modules malfunction when damaged, fall off the mech during the combat and revisit their movement and armor.

Then I'll consider making it a bit more clear in UI or by other means. 
Right know I myself know nothing about it. Just avoided it in this PR.

MTandi committed:
hydraulic clamp force opens airlocks

SyncIt21 commented:
Since you are working on Mechs as a whole could you find the time to fix
#73051
#54688

MTandi committed:
step energy draing and gygax overclocking

MTandi commented:
> Since you are working on Mechs as a whole could you find the time to fix #73051 #54688

Couldn't reproduce storm issue on my branch (guess cabin changes automatically fixed it)
Reworked mecha energy movement and Gygax overclocking specifically.

Also fixed another bug. It seems like there was no point in adding better parts to syndicate mechs because they were not applying their values on Initialize.

MTandi committed:
capacitor reduces light power usage

MTandi committed:
capacitors reduce mele energy drain

MTandi committed:
internal damage fix and overclocking in ui

MTandi committed:
var name

MTandi committed:
Show percentage in UI

MTandi committed:
no zero

MTandi committed:
crowbar fix

MTandi committed:
more energy drain for these

MTandi committed:
integrity and battery level in UI

SyncIt21 commented:
There seems to be only a few mech related issues open. Rather than linking them here just type "Mecha" in the issue search bar and you pick what issues can be solved if you're intrested that is
![Screenshot (283)](https://github.com/tgstation/tgstation/assets/110812394/65d7c109-aaee-422e-ad2a-3cd8dab3581e)
 interested that is

Sealed101 commented:
this shouldn't be a blanket mecha fixing PR bro, the issues can wait for a separate PR

also most of mech issues are in [this](https://github.com/tgstation/tgstation/milestone/69) milestone with some exceptions

SyncIt21 commented:
> this shouldn't be a blanket mecha fixing PR bro, the issues can wait for a separate PR

You right which is why i said 

> if you're interested

Just something to think about in future PRs then that's all

MTandi commented:
Just review this one please, I'll take a look at others later. Don't want to drag this feature set.

SyncIt21 commented on code/modules/research/designs/mechfabricator_designs.dm:
```diff
@@ -957,6 +957,60 @@
+	build_path = /obj/item/mecha_parts/mecha_equipment/radio
+	materials = list(/datum/material/iron=SHEET_MATERIAL_AMOUNT*2.5)
+	construction_time = 100
+	category = list(
+		RND_CATEGORY_MECHFAB_EQUIPMENT,
+		RND_CATEGORY_MECHFAB_RIPLEY,
+		RND_CATEGORY_MECHFAB_GYGAX,
+		RND_CATEGORY_MECHFAB_DURAND,
+		RND_CATEGORY_MECHFAB_HONK,
+		RND_CATEGORY_MECHFAB_PHAZON,
+		RND_CATEGORY_MECHFAB_CLARKE
+	)
+	category = list(
+		RND_CATEGORY_MECHFAB_EQUIPMENT + RND_SUBCATEGORY_MECHFAB_EQUIPMENT_MINING,
+		RND_CATEGORY_MECHFAB_RIPLEY + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT,
+		RND_CATEGORY_MECHFAB_GYGAX + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT,
+		RND_CATEGORY_MECHFAB_DURAND + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT,
+		RND_CATEGORY_MECHFAB_HONK + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT,
+		RND_CATEGORY_MECHFAB_PHAZON + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT,
+		RND_CATEGORY_MECHFAB_CLARKE + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT
+	)
```
Why is `category` list initialized twice? The 2nd one should be sufficient.

SyncIt21 commented on code/modules/research/designs/mechfabricator_designs.dm:
```diff
@@ -957,6 +957,60 @@
+	build_path = /obj/item/mecha_parts/mecha_equipment/air_tank
+	materials = list(/datum/material/iron=SHEET_MATERIAL_AMOUNT*5)
+	construction_time = 100
+	category = list(
+		RND_CATEGORY_MECHFAB_EQUIPMENT,
+		RND_CATEGORY_MECHFAB_RIPLEY,
+		RND_CATEGORY_MECHFAB_GYGAX,
+		RND_CATEGORY_MECHFAB_DURAND,
+		RND_CATEGORY_MECHFAB_HONK,
+		RND_CATEGORY_MECHFAB_PHAZON,
+		RND_CATEGORY_MECHFAB_CLARKE
+	)
+	category = list(
+		RND_CATEGORY_MECHFAB_EQUIPMENT + RND_SUBCATEGORY_MECHFAB_EQUIPMENT_MINING,
+		RND_CATEGORY_MECHFAB_RIPLEY + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT,
+		RND_CATEGORY_MECHFAB_GYGAX + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT,
+		RND_CATEGORY_MECHFAB_DURAND + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT,
+		RND_CATEGORY_MECHFAB_HONK + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT,
+		RND_CATEGORY_MECHFAB_PHAZON + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT,
+		RND_CATEGORY_MECHFAB_CLARKE + RND_SUBCATEGORY_MECHFAB_SUPPORTED_EQUIPMENT
+	)
```
Same `category` list is initialized. twice

SyncIt21 replied:
Yeah go ahead. The 2nd one should be fine

SyncIt21 commented on code/modules/vehicles/mecha/_mecha.dm:
```diff
@@ -71,19 +68,15 @@
-	var/internal_tank_valve = ONE_ATMOSPHERE
-	///The internal air tank obj of the mech
-	var/obj/machinery/portable_atmospherics/canister/air/internal_tank
+	///Whether the pilot is hidden from the outside viewers and whether the cabin can be sealed to be aritight
```
```suggestion
	///Whether the pilot is hidden from the outside viewers and whether the cabin can be sealed to be airtight
```

SyncIt21 reviewed (commented):
Few nit picks and 1 performance request

MTandi committed:
Update code/modules/vehicles/mecha/_mecha.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

MTandi commented on code/modules/vehicles/mecha/_mecha.dm:
```diff
@@ -53,12 +51,11 @@
-	///Whether the mechs maintenance protocols are on or off
-	var/construction_state = MECHA_LOCKED
+	/// Keeps track of the mech's servo motor
+	var/obj/item/stock_parts/servo/servo
```
I need them to have integrity for future combat changes. I never seen more than a few mechs in a single shift.

SyncIt21 replied:
Ah. You can resolve this then

SyncIt21 commented on code/modules/vehicles/mecha/equipment/mecha_equipment.dm:
```diff
@@ -54,10 +60,11 @@
 	. = ..()
```
immediatly in the next line check this return value and return if true i.e.
```
. = ..()
if(.)
   return
```

We want this client side code to be executed only if all the conditions are met. for example https://github.com/tgstation/tgstation/blob/382b8c12c1d0cd9cae2bd3014f72efba0c1928f8/code/game/objects/items/rcd/RCD.dm#L461-L464. 

One time i forgot to do the early return and ghosts could suddenly mess with the rcd. Just an example

MTandi replied:
This sounds like a feature 😄

SyncIt21 commented on code/modules/vehicles/mecha/equipment/tools/air_tank.dm:
```diff
@@ -0,0 +1,160 @@
+	)
+
+/obj/item/mecha_parts/mecha_equipment/air_tank/ui_act(action, list/params)
+	. = ..()
```
Same early return

SyncIt21 commented on code/modules/vehicles/mecha/equipment/tools/other_tools.dm:
```diff
@@ -299,19 +301,19 @@
 	)
 
 /obj/item/mecha_parts/mecha_equipment/generator/ui_act(action, list/params)
 	. = ..()
```
Ditto

MTandi committed:
double category definitions

MTandi committed:
Merge branch 'mechui' of github.com:MTandi/tgstation into mechui

Sealed101 commented on code/modules/vehicles/mecha/mecha_defense.dm:
```diff
@@ -302,27 +321,38 @@
+	if(!(mecha_flags & PANEL_OPEN) && LAZYLEN(occupants))
+		for(var/mob/occupant as anything in occupants)
+			occupant.show_message(
+				span_userdanger("[user] is trying to open maintenance panel of [src]!"), MSG_VISUAL,
+				span_userdanger("You hear someone trying to open maintenance panel of [src]!"), MSG_AUDIBLE,
+			)
+		visible_message(span_danger("[user] is trying to open maintenance panel of [src]!"))
+		if(!do_after(user, 5 SECONDS, src))
+			return
+		for(var/mob/occupant as anything in occupants)
+			occupant.show_message(
+				span_userdanger("[user] has opened maintenance panel of [src]!"), MSG_VISUAL,
+				span_userdanger("You hear someone opening maintenance panel of [src]!"), MSG_AUDIBLE,
+			)
+		visible_message(span_danger("[user] has opened maintenance panel of [src]!"))
```
```suggestion
				span_userdanger("[user] is trying to open the maintenance panel of [src]!"), MSG_VISUAL,
				span_userdanger("You hear someone trying to open the maintenance panel of [src]!"), MSG_AUDIBLE,
			)
		visible_message(span_danger("[user] is trying to open the maintenance panel of [src]!"))
		if(!do_after(user, 5 SECONDS, src))
			return
		for(var/mob/occupant as anything in occupants)
			occupant.show_message(
				span_userdanger("[user] has opened the maintenance panel of [src]!"), MSG_VISUAL,
				span_userdanger("You hear someone opening the maintenance panel of [src]!"), MSG_AUDIBLE,
			)
		visible_message(span_danger("[user] has opened the maintenance panel of [src]!"))
```

SyncIt21 reviewed (commented):
Some `ui_act()` changes. Make sure to check all over for this pattern

Sealed101 commented on code/modules/vehicles/mecha/mecha_defense.dm:
```diff
@@ -331,15 +361,37 @@
+		var/obj/item/stock_parts/part_to_remove = tgui_input_list(user, "Which part to remove?", "Part Removal", stock_parts)
+		if(!(locate(part_to_remove) in contents))
+			return
+		usr.put_in_hands(part_to_remove)
```
```suggestion
		user.put_in_hands(part_to_remove)
```

MTandi committed:
if dot then return

MTandi committed:
here too

Sealed101 reviewed (commented):
some grammar nits and whatnot

MTandi committed:
Update code/modules/vehicles/mecha/_mecha.dm

Co-authored-by: Sealed101 <cool.bullseye@yandex.ru>

MTandi committed:
Update code/modules/vehicles/mecha/_mecha.dm

Co-authored-by: Sealed101 <cool.bullseye@yandex.ru>

MTandi committed:
Update code/modules/vehicles/mecha/mecha_defense.dm

Co-authored-by: Sealed101 <cool.bullseye@yandex.ru>

MTandi committed:
grammar, user, sealed

MTandi committed:
Merge branch 'mechui' of github.com:MTandi/tgstation into mechui

MTandi committed:
oops

MTandi committed:
oops2

SyncIt21 commented on code/modules/vehicles/mecha/equipment/weapons/weapons.dm:
```diff
@@ -477,7 +478,8 @@
 	.["mode_label"] = "Honk Severiy"
 
 /obj/item/mecha_parts/mecha_equipment/weapon/ballistic/launcher/punching_glove/ui_act(action, list/params)
-	. = ..()
```
Ditto

MTandi committed:
runtime fix

SyncIt21 commented on code/modules/vehicles/mecha/mecha_defense.dm:
```diff
@@ -219,17 +219,22 @@
 
 	if(istype(W, /obj/item/stock_parts/cell))
-		if(construction_state == MECHA_OPEN_HATCH)
+		if(mecha_flags & PANEL_OPEN)
```
This if condition is repeated a lot of times, Try writing an early return for it for e.g.

```
if(istype(W, obj/item/stock_parts))
  //If we are trying to insert a stock part, don't bother if the panel isn't open
  if(!(mecha_flags & PANEL_OPEN))
     return
  .
  //check for particular type of stock part and do the regular stuff
```
Then you will have less repeated code

MTandi replied:
That will make it impossible to attach with anything while not in combat mode and the panel is closed. And everything in `return ..()` won't work if the panel is closed. I can reduce the nesting, but the check has to stay there, I believe.

SyncIt21 replied:
Whatever you can do to reduce the code then,  go ahead

MTandi replied:
Moved into a new proc try_insert_part and checking the panel inside

SyncIt21 replied:
Looks good. You can resolve this

SyncIt21 commented on code/modules/vehicles/mecha/mecha_defense.dm:
```diff
@@ -302,27 +321,38 @@
 
 /obj/vehicle/sealed/mecha/examine(mob/user)
 	.=..()
```
```suggestion
	. = ..()
```
space

SyncIt21 commented on code/modules/vehicles/mecha/mecha_defense.dm:
```diff
@@ -302,27 +321,38 @@
+		visible_message(span_danger("[user] has opened the maintenance panel of [src]!"))
+
+	mecha_flags ^= PANEL_OPEN
+	if(mecha_flags & PANEL_OPEN)
+		balloon_alert(user, "panel open")
+	else
+		balloon_alert(user, "panel closed")
```
```suggestion
		balloon_alert(user, "panel [(mecha_flags & PANEL_OPEN) ? "open" : "closed"]")
```
One liner looks nicer

SyncIt21 commented on code/modules/vehicles/mecha/mecha_defense.dm:
```diff
@@ -302,27 +321,38 @@
+	else
+		balloon_alert(user, "panel closed")
+	tool.play_tool_sound(src)
+	return
```
```suggestion
```
we already reached the end of this proc. Not required

SyncIt21 commented on code/modules/vehicles/mecha/mecha_defense.dm:
```diff
@@ -219,17 +219,22 @@
+		to_chat(user, span_warning("ID lock [mecha_flags & ID_LOCK_ON ? "enabled" : "disabled"]."))
+		return
+
+	if(is_wire_tool(W) && (mecha_flags & PANEL_OPEN))
```
no single var name `W` put `weapon` or something

SyncIt21 reviewed (commented):
Getting there.

MTandi committed:
construction fix

MTandi committed:
feedback

MTandi committed:
fixes

MTandi committed:
CheckParts

MTandi committed:
Merge branch 'master' into mechui

MTandi committed:
cleaner try_insert_part proc

MTandi committed:
ripley constructrion icon fix

MTandi committed:
initial parts cleaner

SyncIt21 commented on code/modules/vehicles/mecha/_mecha.dm:
```diff
@@ -485,69 +490,75 @@
+	if(internal_damage & MECHA_INT_SHORT_CIRCUIT && get_charge())
+		spark_system.start()
+		use_power(min(10 * seconds_per_tick, cell.charge))
+		cell.maxcharge -= min(10 * seconds_per_tick, cell.maxcharge)
```
you shouldn't be changing the cells `maxcharge` this value is constant based on the cell type. Instead call `/obj/item/stock_parts/cell/use()` It does blackbox recording and also checks if the cell is rigged

https://github.com/tgstation/tgstation/blob/f9ca87b95976476540f70128809ad837967d9f6b/code/modules/power/cell.dm#L141-L150

SyncIt21 replied:
Check elsewhere and make the same changes

MTandi replied:
It's the previous behavior of this type of internal damage. I assume the author wanted to make the cell capacity to decay on purpose.

SyncIt21 replied:
Actually never mind. `use_power()` already calls this proc. But i'm not sure why the cells max charge should be decreased over time. I guess that's a feature.

SyncIt21 replied:
Yeah you right. You can resolve this

MTandi replied:
I added the maxcharge display to the UI so that the pilot can see the battery degrading over time. It was unclear when the power was going down together with maxcharge and percentage barely changed.

MTandi commented:
There is a bug related to stock parts loading on mech construction completion / initialization. Marking as draft until I resolve it.

SyncIt21 commented:
> There is a bug related to stock parts loading on mech construction completion / initialization. Marking as draft until I resolve it.

You could create a map load variant which has a flag say `init_stock_parts = TRUE` and you call `CheckParts()` , `add_servo()` etc only when this flag is true. By default this flag would be false of course

MTandi committed:
initialzie_parts

MTandi committed:
parts stuff

MTandi committed:
parts fixed

Jacquerel commented on code/modules/vehicles/mecha/_mecha.dm:
```diff
@@ -199,30 +200,29 @@
 
-/obj/item/radio/mech //this has to go somewhere
-	subspace_transmission = TRUE
+/obj/vehicle/sealed/mecha/New(loc, built_manually, ...)
+	src.built_manually = built_manually
+	return ..()
```
Can this not also be handled in Initialize?
It seems like if it was, you wouldn't even need to hold a reference to the variable as it is also only used in `Initialize`.

Jacquerel commented on code/modules/vehicles/mecha/equipment/tools/air_tank.dm:
```diff
@@ -0,0 +1,162 @@
+			var/new_pressure = text2num(params["new_pressure"])
+			tank_pump_pressure = clamp(round(new_pressure), PUMP_MIN_PRESSURE, min(PUMP_MAX_PRESSURE, internal_tank.maximum_pressure))
+			return TRUE
+		else
+			return FALSE
```
```suggestion
	return FALSE
```

Jacquerel commented on code/modules/vehicles/mecha/equipment/tools/radio.dm:
```diff
@@ -0,0 +1,47 @@
+	desc = "A basic component of every vehicle."
+	icon_state = "mecha_radio"
+	equipment_slot = MECHA_UTILITY
+	/// Typepath of radio item
```
Comment isn't true, this is storing a reference not a typepath.

Jacquerel commented on code/modules/vehicles/mecha/equipment/tools/radio.dm:
```diff
@@ -0,0 +1,47 @@
+
+/obj/item/mecha_parts/mecha_equipment/radio/Initialize(mapload)
+	. = ..()
+	radio = new(src)
```
Maybe register the `COMSIG_QDELETED` signal on the radio too so our reference is cleared, just so that it does not hard delete in case some weird circumstance causes it to be deleted.

Jacquerel commented on code/modules/vehicles/mecha/equipment/tools/radio.dm:
```diff
@@ -0,0 +1,47 @@
+			var/new_frequency = text2num(params["new_frequency"])
+			radio.set_frequency(sanitize_frequency(new_frequency, radio.freerange, radio.syndie))
+			return TRUE
+		else
+			return FALSE
```
```suggestion
	return FALSE
```

Jacquerel commented on code/modules/vehicles/mecha/mecha_defense.dm:
```diff
@@ -204,78 +204,103 @@
-			to_chat(user, span_notice("[src]-[W] interface initialized successfully."))
+	if(istype(weapon, /obj/item/mmi))
+		if(mmi_move_inside(weapon,user))
+			to_chat(user, span_notice("[src]-[weapon] interface initialized successfully."))
```
A lot of these could probably be balloon alerts 🤔 
I am not requiring this as a blocker on merging though.

Jacquerel commented on code/modules/vehicles/mecha/mecha_defense.dm:
```diff
@@ -204,78 +204,103 @@
+			if(!user.transferItemToLoc(weapon, src, silent = FALSE))
+				return
+			scanmod = weapon
+			to_chat(user, span_notice("You install the scanning module."))
```
Same here wrt balloon alerts. This isn't important information which needs to be persistently logged for the user.

MTandi committed:
cleaner lights code

MTandi committed:
Update code/modules/vehicles/mecha/equipment/tools/air_tank.dm

Co-authored-by: Jacquerel <hnevard@gmail.com>

MTandi committed:
Update code/modules/vehicles/mecha/equipment/tools/radio.dm

Co-authored-by: Jacquerel <hnevard@gmail.com>

MTandi committed:
feedback

MTandi committed:
Merge branch 'mechui' of github.com:MTandi/tgstation into mechui

MTandi committed:
feedback

MTandi committed:
lint

MTandi committed:
ID to id

MTandi committed:
infinite qdel loop
